src/globus_cli/login_manager/tokenstore.py

```diff
@@ -10,15 +11,218 @@
+    ) -> dict[str, t.Any]: ...
+
+    @t.overload
+    def read_well_known_config(
+        self,
+        name: t.Literal[
+            "auth_client_data", "auth_user_data", "scope_contract_versions"
+        ],
+        *,
+        allow_null: bool = True,
+    ) -> dict[str, t.Any] | None: ...
+
+    def read_well_known_config(
+        self,
+        name: t.Literal[
+            "auth_client_data", "auth_user_data", "scope_contract_versions"
+        ],
+        *,
+        allow_null: bool = True,
+    ) -> dict[str, t.Any] | None:
+        if self._adapter is None:
+            raise RuntimeError("cannot read well-known config after closing storage")
+        data = self._adapter.read_config(name)
+        if not allow_null and data is None:
+            if name == "auth_user_data":
+                alias = "Identity Info"
+            else:
+                alias = name
+            raise RuntimeError(
+                f"{alias} was unexpectedly not visible in storage. "
+                "A new login should fix the issue. "
+                "Consider using `globus login --force`"
+            )
+        return data
+
+    def remove_well_known_config(
+        self,
+        name: t.Literal[
+            "auth_client_data", "auth_user_data", "scope_contract_versions"
+        ],
+    ) -> None:
+        if self._adapter is None:
+            raise RuntimeError("cannot remove well-known config after closing storage")
+        self._adapter.remove_config(name)
+
+    #
+    # APIs which match the tokenstorage itself
+    #
+
+    def get_token_data(self, resource_server: str) -> dict[str, t.Any] | None:
+        if self._adapter is None:
+            raise RuntimeError("cannot get token data after closing storage")
+        return self._adapter.get_token_data(resource_server)
+
+    def get_by_resource_server(self) -> dict[str, dict[str, t.Any]]:
+        if self._adapter is None:
+            raise RuntimeError("cannot get token data after closing storage")
+        return self._adapter.get_by_resource_server()
+
+    def store(self, token_response: globus_sdk.OAuthTokenResponse) -> None:
+        if self._adapter is None:
+            raise RuntimeError("cannot store token data after closing storage")
+        self._adapter.store(token_response)
+        # store contract versions for all of the tokens which were acquired
+        # this could overwrite data from another CLI version *earlier or later* than
+        # the current one
+        #
+        # in the case that the old data was from a prior version, this makes sense
+        # because we have added new constraints or behaviors
+        #
+        # if the data was from a *newer* CLI version than what we are currently
+        # running we can't really know with certainty that "downgrading" the version
+        # numbers is correct, but because we can't know we need to just do our best
+        # to indicate that the tokens in storage may have lost capabilities
```

Comment:
How are we indicating that the version was downgraded? Looks like we're just overriding with whatever the current version is no matter what.

Reply:
I think this comment is correct. `store()` being called means we're writing tokens to disk.

We are, as you say, overwriting any previous version marker (top of the comment calls this out as well). The comment indicates that this could "downgrade" tokens -- i.e. store tokens which are strictly weaker than the ones which were available on disk, which would presumably have a higher version number in most cases -- and it would not be recoverable information.

In fact, we've only used the contract version mechanism once, at the time we introduced it, so I wonder if it's a less general problem than we believed at the time, or if we've simply not been forced to reckon with scope changes in other services yet.